src/MSALWrapper/AuthFlow/IntegratedWindowsAuthentication.cs

```diff
@@ -53,51 +52,46 @@ public IntegratedWindowsAuthentication(ILogger logger, Guid clientId, Guid tenan
-            this.logger.LogDebug($"Using cached account '{account?.Username}'");
+            IAccount account = await this.pcaWrapper.TryToGetCachedAccountAsync(this.preferredDomain);
             TokenResult tokenResult = null;
 
             try
             {
-                try
-                {
-                    tokenResult = await TaskExecutor.CompleteWithin(
-                        this.logger,
-                        this.integratedWindowsAuthTimeout,
-                        "Get Token Silent",
-                        (cancellationToken) => this.pcaWrapper.GetTokenSilentAsync(this.scopes, account, cancellationToken),
-                        this.errors)
-                        .ConfigureAwait(false);
-                    tokenResult.SetSilent();
-                    if (tokenResult == null)
-                    {
-                        this.errors.Add(new NullTokenResultException("IWA Get Token Silent returned null.(Not expected)"));
-                    }
-                }
-                catch (MsalUiRequiredException ex)
+                tokenResult = await CachedAuth.GetTokenAsync(
+                    this.logger,
+                    this.scopes,
+                    account,
+                    this.pcaWrapper,
+                    this.errors);
+
+                if (tokenResult != null)
                 {
-                    this.errors.Add(ex);
-                    this.logger.LogDebug("Cached auth failed.");
-                    this.logger.LogDebug(ex.Message);
-                    tokenResult = await TaskExecutor.CompleteWithin(
-                                  this.logger,
-                                  this.integratedWindowsAuthTimeout,
-                                  "Get Token Integrated Windows Authentication",
-                                  (cancellationToken) => this.pcaWrapper.GetTokenIntegratedWindowsAuthenticationAsync(this.scopes, cancellationToken),
-                                  this.errors)
-                                  .ConfigureAwait(false);
-                    tokenResult.SetSilent();
+                    return (tokenResult, this.errors);
                 }
+
+                Func<CancellationToken, Task<TokenResult>> iwaAuth = (cancellationToken) =>
+                    this.pcaWrapper.GetTokenIntegratedWindowsAuthenticationAsync(this.scopes, cancellationToken);
+
+                tokenResult = await TaskExecutor.CompleteWithin(
+                    this.logger,
+                    this.integratedWindowsAuthTimeout,
+                    "Integrated Windows Authentication",
+                    iwaAuth,
+                    this.errors).ConfigureAwait(false);
+
+                // If IWA worked, it was 100% silent.
+                tokenResult.SetSilent();
             }
             catch (MsalUiRequiredException ex)
             {
                 this.errors.Add(ex);
+                this.logger.LogDebug(ex.Message);

```

Comment:
Consider adding log assertions in tests.

Reply:
Will do post test refactor

Reply:
Resolving with #282